src/main/scala/chisel3/util/LoadMemoryTransform.scala

```diff
@@ -0,0 +1,174 @@
+// See LICENSE for license details.
+
+package chisel3.util
+
+import chisel3.MemBase
+import chisel3.core.{ChiselAnnotation, RunFirrtlTransform, annotate}
+import chisel3.internal.{Builder, InstanceId}
+import firrtl.annotations.{MemoryLoadFileType, _}
+import firrtl.ir.{Module => _, _}
+import firrtl.transforms.BlackBoxInlineAnno
+import firrtl.{AnnotationSeq, CircuitForm, CircuitState, EmitCircuitAnnotation, LowForm, Transform, VerilogEmitter}
+
+import scala.collection.mutable
+
+/**
+  * chisel implementation for load memory
+  * @param target        memory to load
+  * @param fileName      name of input file
+  * @param hexOrBinary   use $readmemh or $readmemb
+  */
+case class ChiselLoadMemoryAnnotation(
+  target:      InstanceId,
+  fileName:    String,
+  hexOrBinary: MemoryLoadFileType.FileType = MemoryLoadFileType.Hex
+)
+  extends ChiselAnnotation with RunFirrtlTransform {
+
+  if(fileName.isEmpty) {
+    Builder.warning(
+      s"""LoadMemory from file annotations file empty file name"""
+    )
+  }
+
+  def transformClass: Class[LoadMemoryTransform] = classOf[LoadMemoryTransform]
+
+  def toFirrtl: LoadMemoryAnnotation = {
+    LoadMemoryAnnotation(target.toNamed.asInstanceOf[ComponentName], fileName, hexOrBinary)
+  }
+}
+
+
+object loadMemoryFromFile {
+  def apply(
+    memory: MemBase[_],
+    fileName: String,
+    hexOrBinary: MemoryLoadFileType.FileType = MemoryLoadFileType.Hex
+  ): Unit = {
+    annotate(ChiselLoadMemoryAnnotation(memory, fileName))
+  }
+}
+
+/**
+  * This transform only is activated if verilog is being generated
+  * (determined by presence of the proper emit annotation)
+  * when activated it creates additional verilog files that contain
+  * modules bound to the modules that contain an initializable memory
+  *
+  * Currently the only non-verilog based simulation that can support loading
+  * memory from a file is treadle but it does not need this transform
+  * to do that.
+  */
+//noinspection ScalaStyle
+class LoadMemoryTransform extends Transform {
+  def inputForm: CircuitForm  = LowForm
+  def outputForm: CircuitForm = LowForm
+
+  private var memoryCounter: Int = -1
+
+  private val bindModules: mutable.ArrayBuffer[BlackBoxInlineAnno] = new mutable.ArrayBuffer()
+
+  private val verilogEmitter:    VerilogEmitter = new VerilogEmitter
+
+  /**
+    * run the pass
+    * @param circuit the circuit
+    * @param annotations all the annotations
+    * @return
+    */
+  def run(circuit: Circuit, annotations: AnnotationSeq): Circuit = {
+
+    val memoryAnnotations = {
+      annotations.collect{ case m: LoadMemoryAnnotation => m }.map { ma => ma.target.serialize -> ma }.toMap
```

Comment:
Note: I didn't explicitly specify it, but the above stays as `Map[String, LoadMemoryAnnotation]`. No need for modifying later code in the transform.

Reply:
I have switched the code to use #groupBy. I am not sure that it can fail given the way module deduping and annotation management works.  But code is cleaner

Reply:
done